# Data Visualization (2018/19)

## Solutions for Assignment 2 - Exploratory Data Analysis

Presented by Group 25: 
- Ganesh
- Chaitra

Date: 22.11.2018, 22:00

Hints: 
- Double click on a cell to edit it.
- Enter the Markdown mode in the drop-down menu above (it may say Code for new cells).
- Run a cell by pressing CTRL+Enter
- Markdown docu: http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html
- Markdown cheatsheet: http://datascience.ibm.com/blog/markdown-for-jupyter-notebooks-cheatsheet/

### Exercise 1: Choose a data set

We will use the Python Data Analysis Library (Pandas) to work with the data. Pandas provides a rich feature set for working with tabular data including data structures and analysis tools.

A 10 minute introduction to pandas with all neccessary functions can be found here: http://pandas.pydata.org/pandas-docs/stable/10min.html (Quickly skim over the available functionality. You don't have to learn it yet.)

In [1]:
# import the pandas library and call it pd for further usage
import pandas as pd

#### Sample code to load the Titanic data and get basic statistics

In [2]:
# read the csv file into a pandas dataframe and print the first lines of the table
filename = "./data/titanic3.csv"
df = pd.read_csv( filename, header=0 )
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [3]:
df.dtypes

pclass         int64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object

In [4]:
df.describe(include="all")

pclass     survived                  name   sex          age  \
count   1309.000000  1309.000000                  1309  1309  1046.000000   
unique          NaN          NaN                  1307     2          NaN   
top             NaN          NaN  Connolly, Miss. Kate  male          NaN   
freq            NaN          NaN                     2   843          NaN   
mean       2.294882     0.381971                   NaN   NaN    29.881138   
std        0.837836     0.486055                   NaN   NaN    14.413493   
min        1.000000     0.000000                   NaN   NaN     0.170000   
25%        2.000000     0.000000                   NaN   NaN    21.000000   
50%        3.000000     0.000000                   NaN   NaN    28.000000   
75%        3.000000     1.000000                   NaN   NaN    39.000000   
max        3.000000     1.000000                   NaN   NaN    80.000000   

              sibsp        parch    ticket         fare        cabin embarked  \
count   1309.000000  1309.000000      1309  1308.000000          295     1307   
unique          NaN          NaN       929          NaN          186        3   
top             NaN          NaN  CA. 2343          NaN  C23 C25 C27        S   
freq            NaN          NaN        11          NaN            6      914   
mean       0.498854     0.385027       NaN    33.295479          NaN      NaN   
std        1.041658     0.865560       NaN    51.758668          NaN      NaN   
min        0.000000     0.000000       NaN     0.000000          NaN      NaN   
25%        0.000000     0.000000       NaN     7.895800          NaN      NaN   
50%        0.000000     0.000000       NaN    14.454200          NaN      NaN   
75%        1.000000     0.000000       NaN    31.275000          NaN      NaN   
max        8.000000     9.000000       NaN   512.329200          NaN      NaN   

       boat        body     home.dest  
count   486  121.000000           745  
unique   27         NaN           369  
top      13         NaN  New York, NY  
freq     39         NaN            64  
mean    NaN  160.809917           NaN  
std     NaN   97.696922           NaN  
min     NaN    1.000000           NaN  
25%     NaN   72.000000           NaN  
50%     NaN  155.000000           NaN  
75%     NaN  256.000000           NaN  
max     NaN  328.000000           NaN

#### Your selection:
Selected dataset: Titanic Dataset

### Exercise 2: Data queries

1. What is the total number of people on board based on gender? 
2. What is the co-relation between the survivors and casualties based on age (greater than 50 vs less than equal to 50)?
3. Among the casualities, what is the count of marked Body Identification Number vs unmarked Body Identification Number?
4. How many tickets were purchased in passenger classes 1, 2 and 3 ?
5. What are the top destinations traveller wished to visit?
6. What is the total count of people at different ports of embarkation

### Exercise 3: Data analysis

Here is sample code to pass your data from the pandas data frame to bokeh and render a graphic inside the notebook.

In [5]:
# import bokeh 
from bokeh.plotting import figure, show, output_file
from bokeh.models import ColumnDataSource, FactorRange
from bokeh.transform import dodge, factor_cmap, cumsum
from bokeh.palettes import Spectral6, Category20c, Spectral3
from bokeh.core.properties import value
import math

# tell bokeh to show the figures in the notebook
from bokeh.io import output_notebook
from bokeh.layouts import row, column
output_notebook()
output_file("Assignment2_Ganesh-Chaitra.html")

Loading BokehJS ...

1. What is the total number of people on board based on gender? 
Ans: Among the people on-board, 843 were male and 466 were female. It can be observed that, number of male travellers were more compared to female.

In [6]:
#Question 1
total_people = pd.DataFrame({'count': df.groupby(["sex"]).size()}).reset_index()

male_female = total_people['sex'].tolist()
count = total_people['count'].tolist()

source_ques1 = ColumnDataSource(data=dict(male_female=male_female, count=count))

2. What is the co-relation between the survivors and casualties based on age (greater than 50 vs less than equal to 50)? Ans: Among the people on-board, there were 619 casualities and 427 survivors. It can be seen from the graph that there were more casualities in both sections of age than the survivors. However, data concerning age for few people was missing. Hence, this can further be investigated in getting a more accurate result. 

In [7]:
#Question 2
people_onboard = pd.DataFrame({'age<=50': df[df['age'] <= 50].groupby('survived')['age'].count(),
                              'age>50': df[df['age'] > 50].groupby('survived')['age'].count()}).reset_index()
#people_onboard
ageLabel = ['age<=50', 'age>50']
surviveLabel = ['Casualities','Survivors']

data_ques2 = people_onboard.to_dict(orient='list')
label_list = [(data_survived, data_age) for data_survived in surviveLabel for data_age in ageLabel]
count_data_list = sum(zip(data_ques2['age<=50'], data_ques2['age>50']), ())

source_ques2 = ColumnDataSource(data=dict(label_list=label_list, count_data_list=count_data_list))


3. Among the casualities, what is the count of marked Body Identification Number vs unmarked Body Identification Number (BIN)? Ans: Among the casualities, there are 688 with unmarked body identification number which is much more than the marked ones (121). This data shows how difficult it can be to identify a person without a BIN.

In [8]:
#Question 3
#create a new df consisting data of only casualities. BIN-Body Identification Number
df_casualities = df[df['survived'] == 0]
df_casualities_count = pd.DataFrame({'marked BIN': df_casualities[df_casualities['body'].notnull()].groupby('survived')['body'].count(),
                                     'unmarked BIN': sum(pd.isnull(df_casualities['body']))}).reset_index()
data_ques3 = df_casualities_count.to_dict(orient='list')

survived = ['Casualities']
BIN_label = ['marked BIN','unmarked BIN']
colors = ["#ff0000", "#7ec0ee"]


4. How many tickets were purchased in passenger classes 1 (upper), 2 (middle)and 3 (lower)? Ans: 709 tickets were purchased in Passenger class 3, 277 in 2 and 323 in 1. It can be observed that tickets purchased in lower class were more than upper and middle class combined.

In [9]:
#Question 4
tickets_pclass = df.groupby('pclass')['ticket'].describe().rename(columns={'count': 'tickets_purchased'}).reset_index()
#tickets_pclass.head()
#data_ques4 = tickets_pclass.to_dict(orient='list')
pclass = tickets_pclass['pclass'].tolist()
tickets_purchased = tickets_pclass['tickets_purchased'].tolist()
pclass = ['Upper', 'Middle', 'Lower']

source_ques4 = ColumnDataSource(data=dict(pclass=pclass, tickets_purchased=tickets_purchased))


5. What are the top destinations traveller wished to visit? Ans: As seen from the visualization, top 20 destinations are the ones listed and New york tops the chart. While 64 wished to go to New York, only 4 wished to go to Ruotsinphytaa.

In [10]:
#Question 5

top_destination_list = df.groupby('home.dest').size().sort_values(ascending=False)
top_destination_dict = {}
index=0
threshold = 20
for key,count in top_destination_list.items(): 
    if index < threshold:
        top_destination_dict[key] = count
        index +=1
    else: break

data_ques5 = pd.Series(top_destination_dict).reset_index(name='value').rename(columns={'index':'city'})
data_ques5['angle'] = data_ques5['value']/data_ques5['value'].sum() * 2*math.pi
data_ques5['color'] = Category20c[len(top_destination_dict)]

6. What is the total count of people at different ports of embarkation? Ans: From the chart, it can be observed that 914 embarked from Southampton, 123 from Queentstown and 270 from Cherbourg. Total number of people embarked at Southampton is a lot compared to the rest.

In [11]:
#Question 6
embarked_at = df.groupby('embarked')['pclass'].describe().rename(columns={'count': 'embarked_count'}).reset_index()
embarked_at

#['Cherbourg','Queenstown','Southampton']
embarked_label = ['C','Q','S']

source_ques6 = ColumnDataSource(data=dict(embarked=embarked_at['embarked'], embarked_count=embarked_at['embarked_count'], color=Spectral3))


In [12]:
#Start - Question 1
p = figure(plot_height=300, title="Total number of people onboard based on gender.", x_range=male_female, 
           tools="hover", y_range = (0,1000))

p.vbar(source=source_ques1, x='male_female', width=0.3, bottom=0, top='count',
        line_color='white', fill_color=factor_cmap('male_female', palette=Spectral6, factors=male_female))

p.hover.tooltips = [
    ("On-board", "@count")
]


p.xaxis.axis_label = "Males/Females"
p.yaxis.axis_label = "count"
p.xgrid.visible = False

#show(p)
#End
#Start - Question 2
q = figure(x_range=FactorRange(*label_list), plot_height=250, plot_width=800, tools='hover',
           title="Co-relation between the survivors and casualties based on age (greater than 50 vs less than equal to 50).")

q.vbar(x='label_list', top='count_data_list', width=0.5, source=source_ques2, line_color="white", color="pink")

q.hover.tooltips = [
    ("Count", "@count_data_list")
]

q.yaxis.axis_label = "count"

# #show(q)
# #End
# #Start - Question 3
r = figure(y_range=survived, plot_height=250, plot_width=900,
           title="Among the casualities, marked Body Identification Number vs unmarked Body Identification Number.",
           toolbar_location=None, tools="hover", tooltips="$name: @$name")

r.hbar_stack(BIN_label, y='survived', height=0.5, color=colors, source=data_ques3,
             legend=[value(y) for y in BIN_label])

r.xaxis.axis_label = "Body Indentification Number count"

# #show(r)
# #End
# #Start - Question 4
s = figure(plot_height=300, title="Number of tickets purchased on Passenger Class 1, 2 and 3.", x_range=pclass, 
           tools="hover", y_range = (0,1000))

s.vbar(source=source_ques4, x='pclass', width=0.3, bottom=0, top='tickets_purchased',
        line_color='white', fill_color=factor_cmap('pclass', palette=Spectral3, factors=pclass))

s.hover.tooltips = [
    ("Number of tickets purchased", "@tickets_purchased")
]

s.xaxis.axis_label = "Passenger Class"
s.yaxis.axis_label = "Tickets Purchased"

# #show(s)
# #End
# #Start - Question 5
t = figure(plot_height=800, plot_width=1000, title="Top Destinations.", toolbar_location=None,
           tools="hover", tooltips="@city: @value", x_range=(-0.5, 1.0))

t.wedge(x=0, y=1, radius=0.4,
        start_angle=cumsum('angle', include_zero=True), end_angle=cumsum('angle'),
        line_color="white", fill_color='color', legend='city', source=data_ques5)

t.axis.axis_label=None
t.axis.visible=False
t.grid.grid_line_color = None

# #show(t)
# #End
# #Start - Question 6
u = figure(y_range=embarked_label, x_range=(0,1000), plot_height=400, title="Total number of people at different ports of embarkation.",
           toolbar_location=None, tools="hover")

u.hbar(y='embarked', right='embarked_count', height=0.3, color='color', source=source_ques6)

u.hover.tooltips = [
    ("Count:", "@embarked_count")
]
u.yaxis.axis_label = "Embarked at (C-Cherbourg, Q-Queenstown, S-Southampton)"
u.xaxis.axis_label = "Total number of people embarked"

# #show(u)
# #End
show(column(p,q,r,s,t,u))

### Exercise 4: Write a summary

The titanic dataset gives us a brief idea about the overall fleet status. The various aspects of the data includes the ratios of male to female passengers on board, survival status of individual passengers based on a certain age group, their passenger class information and places of destination. The fleet started with huge number of people embarking from Southampton followed by Cherbourg and Queenstown with more male passengers than that of females ones with most of them destined towards Newyork followed by London and so on. The fleet ended up in very devestating state with 619 total casualities and among these only 121 were identified with a BIN. By looking at the data we can also justify that most of the travellers travelled in lower class. The difficulties with the data set was to calculate the co-relation between survivor and casualities as there were lot of NAN's with respect to age. 
